lib/src/main/java/com/diffplug/spotless/generic/PipeStepPair.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.generic;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.diffplug.spotless.FormatterStep;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+public class PipeStepPair {
+	/** The two steps will be named `<name>In` and `<name>Out`. */
+	public static Builder named(String name) {
+		return new Builder(name);
+	}
+
+	public static String defaultToggleName() {
+		return "toggle";
+	}
+
+	public static String defaultToggleOff() {
+		return "spotless:off";
+	}
+
+	public static String defaultToggleOn() {
+		return "spotless:on";
+	}
+
+	public static class Builder {
+		String name;
+		Pattern regex;
+
+		private Builder(String name) {
+			this.name = Objects.requireNonNull(name);
+		}
+
+		/** Defines the opening and closing markers. */
+		public Builder openClose(String open, String close) {
+			return regex(Pattern.quote(open) + "([\\s\\S]*?)" + Pattern.quote(close));
+		}
+
+		/** Defines the pipe via regex. Must have *exactly one* capturing group. */
+		public Builder regex(String regex) {
+			return regex(Pattern.compile(regex));
+		}
+
+		/** Defines the pipe via regex. Must have *exactly one* capturing group. */
+		public Builder regex(Pattern regex) {
+			this.regex = regex;
+			return this;
+		}
+
+		public PipeStepPair buildPair() {
+			return new PipeStepPair(name, regex);
+		}
+	}
+
+	final FormatterStep in, out;
+
+	private PipeStepPair(String name, Pattern pattern) {
+		StateIn stateIn = new StateIn(pattern);
+		StateOut stateOut = new StateOut(stateIn);
+		in = FormatterStep.create(name + "In", stateIn, state -> state::format);
+		out = FormatterStep.create(name + "Out", stateOut, state -> state::format);
+	}
```

Comment:
Minor nit: Consider adding tests that check passing `null` to the `name` and `pattern` parameters of this constructor results in a `NullPointerException`.

Reply:
Added `Objects.requireNonNull` on the name, didn't exhaustively test.